File: ui/src/lib/ipc/record_commands.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  CommandName,
  DefaultFormatDto,
  MonitorMode,
  RecordOffsetDto,
  RecordOffsetSource,
  RecordStartedDto,
  RecordStateDto,
} from "./bindings";

/**
 * S1-04 recording commands: typed `invoke` wrappers (ADR-003) using generated types only. Kept
 * apart from `commands.ts` so the recording slice stays self-contained.
 */

/** The record panel state. */
export async function recordGet(): Promise<RecordStateDto> {
  return invoke<RecordStateDto>("record_get" satisfies CommandName);
}

/** Arms (opens the input, starts the meter) or disarms. */
export async function recordArm(armed: boolean): Promise<RecordStateDto> {
  return invoke<RecordStateDto>("record_arm" satisfies CommandName, { armed });
}

/**
 * Starts a new recording; `replace` = the user confirmed replacing a document with audio.
 * `format` (H-06, the New Recording dialog): the chosen sample rate / bit depth — omitted, Record
 * with no document uses the current default format without showing the dialog (SPEC-002 §2.2).
 */
export async function recordStart(replace: boolean, format?: DefaultFormatDto): Promise<RecordStateDto> {
  return invoke<RecordStateDto>(
    "record_start" satisfies CommandName,
    format ? { replace, format } : { replace },
  );
}

/**
 * T-304 (SPEC-022 §2.2): Record with the current selection — a new recording into an empty
 * document, a punch-in over a non-empty selection, else Insert/Overwrite at the selection start or
 * the cursor.
 */
export async function recordStartAt(selection: [number, number] | null): Promise<RecordStartedDto> {
  return invoke<RecordStartedDto>("record_start_at" satisfies CommandName, { selection });
}

/** Stops the recording (the take becomes the document once committed). */
export async function recordStop(): Promise<RecordStateDto> {
  return invoke<RecordStateDto>("record_stop" satisfies CommandName);
}

/** Sets (and saves) the monitoring mode. */
export async function recordSetMonitor(mode: MonitorMode): Promise<RecordStateDto> {
  return invoke<RecordStateDto>("record_set_monitor" satisfies CommandName, { mode });
}

/**
 * H-07: `count` `(min, max)` buckets of the take being captured, from bucket `startBucket`, as
 * raw `VXPK` bytes (decode with `decodeVxpk`; zero buckets while no take is being captured).
 */
export async function recordPeaksGet(startBucket: number, count: number): Promise<ArrayBuffer> {
  return invoke<ArrayBuffer>("record_peaks_get" satisfies CommandName, { startBucket, count });
}

/** T-304 (SPEC-022 §2.13): the current device setup's recording offset. */
export async function recordOffsetGet(): Promise<RecordOffsetDto> {
  return invoke<RecordOffsetDto>("record_offset_get" satisfies CommandName);
}

/** T-304: stores the offset for the current device setup (clamped to ±500 ms). */
export async function recordOffsetSet(
  offsetMs: number,
  source: RecordOffsetSource,
  confidence: number | null,
): Promise<RecordOffsetDto> {
  return invoke<RecordOffsetDto>("record_offset_set" satisfies CommandName, {
    offsetMs,
    source,
    confidence,
  });
}

/** T-304 (SPEC-022 §2.14): starts a calibration job (`verify`: with the stored offset applied). */
export async function calibrationRun(verify: boolean): Promise<number> {
  return invoke<number>("calibration_run" satisfies CommandName, { verify });
}

/** T-304: aborts the running calibration. */
export async function calibrationCancel(): Promise<void> {
  return invoke<void>("calibration_cancel" satisfies CommandName);
}

File: src-tauri/src/ipc/record_commands.rs
//! Recording commands (S1-04, SPEC-002 §2.1–§2.2, §2.7): arm, record start/stop, monitoring.
//! Thin: each forwards to the [`RecordingService`] (blocking work on `spawn_blocking`).
//!
//! `record_peaks_get` (H-07) is the odd one out: like `document_commands::peaks_get`, it encodes
//! the raw `VXPK` bytes here (ADR-003 §2, CLAUDE.md — bulk IPC data is binary, never JSON floats)
//! since that framing is not `RecordingService`'s business logic to own either.

use tauri::State;
use tauri::ipc::Response;
use vox_engine::record::{LIVE_PEAKS_SPB, LiveTakePeaks};
use vox_project::{VxpkHeader, encode_vxpk};

use crate::ipc::error::IpcError;
use crate::ipc::record_dto::RecordStateDto;
use crate::ipc::{Notice, NoticeLevel};
use crate::recording::RecordingService;
use crate::settings::{DefaultFormatDto, MonitorMode, Settings, SettingsStore};

/// ADR-003 §2's request cap, same as `document_commands::peaks_get`.
const MAX_LIVE_PEAKS_BUCKETS: u32 = 65_536;

async fn blocking<T: Send + 'static>(
    f: impl FnOnce() -> Result<T, IpcError> + Send + 'static,
) -> Result<T, IpcError> {
    tauri::async_runtime::spawn_blocking(f)
        .await
        .map_err(|e| IpcError::internal(e.to_string()))?
}

/// The record panel state.
#[tauri::command]
pub async fn record_get(rec: State<'_, RecordingService>) -> Result<RecordStateDto, IpcError> {
    let rec = rec.inner().clone();
    blocking(move || rec.state()).await
}

/// Arms (opens the input, starts the meter) or disarms the input.
#[tauri::command]
pub async fn record_arm(
    rec: State<'_, RecordingService>,
    armed: bool,
) -> Result<RecordStateDto, IpcError> {
    let rec = rec.inner().clone();
    blocking(move || rec.arm(armed)).await
}

/// Starts a new recording. `replace`: the user confirmed replacing a document that has audio.
/// `format`: the New Recording dialog's chosen sample rate / bit depth (H-06); `None` uses the
/// current default format — Record with no document skips the dialog entirely (SPEC-002 §2.2).
#[tauri::command]
pub async fn record_start(
    rec: State<'_, RecordingService>,
    replace: bool,
    format: Option<DefaultFormatDto>,
) -> Result<RecordStateDto, IpcError> {
    let rec = rec.inner().clone();
    blocking(move || rec.start(replace, format)).await
}

/// Stops the recording; the take becomes the document when it is committed.
#[tauri::command]
pub async fn record_stop(rec: State<'_, RecordingService>) -> Result<RecordStateDto, IpcError> {
    let rec = rec.inner().clone();
    blocking(move || rec.stop()).await
}

/// `count` `(min, max)` buckets of the take being captured, from bucket `start_bucket`, as a
/// binary `VXPK` frame (H-07; `partial: true` always — the take is still growing). Zero buckets
/// at rate 0 while no take is being captured (the UI just keeps showing its last good frame).
#[tauri::command]
pub async fn record_peaks_get(
    rec: State<'_, RecordingService>,
    start_bucket: u32,
    count: u32,
) -> Result<Response, IpcError> {
    let rec = rec.inner().clone();
    let count = count.min(MAX_LIVE_PEAKS_BUCKETS);
    let peaks = blocking(move || Ok(rec.live_peaks(start_bucket, count))).await?;
    let peaks = peaks.unwrap_or(LiveTakePeaks {
        sample_rate_hz: 0,
        len_samples: 0,
        spb: LIVE_PEAKS_SPB,
        buckets: Vec::new(),
    });
    let header = VxpkHeader {
        request_id: 0,
        audio_rev: 0,
        start_sample: u64::from(start_bucket) * u64::from(peaks.spb),
        samples_per_bucket: peaks.spb,
        sample_rate_hz: peaks.sample_rate_hz,
        partial: true,
    };
    Ok(Response::new(encode_vxpk(&header, &peaks.buckets)))
}

/// Sets the monitoring mode and saves it as the preference (SPEC-002 §2.7). The first time
/// monitoring is enabled, a one-time "use headphones" hint is shown.
#[tauri::command]
pub async fn record_set_monitor(
    rec: State<'_, RecordingService>,
    settings: State<'_, SettingsStore>,
    mode: MonitorMode,
) -> Result<RecordStateDto, IpcError> {
    let service = rec.inner().clone();
    let state = blocking(move || service.set_monitor(mode)).await?;
    let before = settings.get();
    let mut saved = before.clone();
    let hint = apply_monitor_pref(&mut saved, mode);
    if saved != before {
        settings.set(saved)?;
    }
    if hint {
        rec.notify(Notice::toast(
            NoticeLevel::Info,
            "notice.monitor.headphones",
        ));
    }
    Ok(state)
}

/// T-107 (SPEC-002 §2.7, AC-12): saves `mode` as the preference; returns whether the one-time
/// headphone hint is due — the first time monitoring is enabled (remembered in the settings).
fn apply_monitor_pref(saved: &mut Settings, mode: MonitorMode) -> bool {
    saved.monitor_mode = mode;
    let hint = mode != MonitorMode::Off && !saved.monitor_hint_shown;
    saved.monitor_hint_shown |= hint;
    hint
}

#[cfg(test)]
mod tests {
    use super::*;

    /// AC-12: the feedback hint appears only the first time monitoring is enabled — not for Off,
    /// not again after it was shown (in any mode), and it survives as a saved setting.
    #[test]
    fn headphone_hint_is_shown_once() {
        let mut s = Settings::default();
        assert!(!apply_monitor_pref(&mut s, MonitorMode::Off));
        assert!(!s.monitor_hint_shown);
        assert!(apply_monitor_pref(&mut s, MonitorMode::Dry));
        assert!(s.monitor_hint_shown);
        assert_eq!(s.monitor_mode, MonitorMode::Dry);
        assert!(!apply_monitor_pref(&mut s, MonitorMode::Off));
        assert!(!apply_monitor_pref(&mut s, MonitorMode::ThroughRack));
        assert_eq!(s.monitor_mode, MonitorMode::ThroughRack);
    }
}
